### PR TITLE
test: coverlet coverage gate + raise non-browser coverage 27% → 42%

### PR DESCRIPTION
**Pull request by AkashBrowserStack**
## What & why

coverlet was present but **unwired**. This enables it (adds `coverlet.msbuild`) with a **41% line gate** in `npm test`, and adds a mock-based test class covering the non-browser logic.

> Honest partial coverage: the live-CLI/Playwright paths can't be unit-tested without a browser+CLI (they run on the CI's full suite). 41% is the deterministic floor, not faked.

## Coverage: 27.38% → **42.36% line** (non-browser, local); higher on the browser CI

New `PercyLogicTest` (33 tests, xunit + Moq `IPage` + reflection for private statics) covers: `ParseWidthsFromOptions`, `TryGetIntFromValue`, `IsResponsiveSnapshotCapture` branches, `CalculateDefaultHeight`, `WaitForReady` disabled paths, `PayloadParser`, `Cache`, `PercyPlaywrightDriver.GetUrl`.

## Gate

- `npm test` now runs `dotnet test /p:CollectCoverage=true /p:ThresholdType=line /p:Threshold=41`.
- Uncovered = live-CLI/Playwright only: `Snapshot`/`GetSerializedDom`/`CaptureResponsiveDom`/`GetResponsiveWidths`/`ProcessFrame` and the driver's `EvaluateSync`/guid-reflection paths — exercised by the existing integration tests on the browser CI.

## Verification

```
dotnet test (non-browser subset) → Passed! 52, Failed: 0; Total Line 42.36% (≥41 gate, exit 0)
```
(Full browser suite + gate verify on this PR's CI — Test runs on pull_request.)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by AkashBrowserStack** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/percy-playwright-dotnet/pull/236  •  **Head:** ba2a941  •  **Reviewers:** stack:code-reviewer

## Summary
Test-only / build-config PR: adds the `coverlet.msbuild` 6.0.2 package, wires a coverlet line-coverage gate into `npm test` (`/p:CollectCoverage=true /p:CoverletOutputFormat=cobertura /p:ThresholdType=line /p:Threshold=41`), and adds `Percy.Test/PercyLogicTest.cs` (374 lines, xUnit) of non-browser pure-logic unit tests reaching private statics in `Percy.cs` via reflection. No production source is modified. Local cobertura reports line-rate 0.4236 (42.4%), clearing the 41 gate (claimed 27%→42%).

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | No secrets; only a localhost CLI port literal (`5338`) in test strings, consistent with existing tests. |
| High | Security | Authentication/authorization checks present | N/A | Test + build-config change only; no auth surface. |
| High | Security | Input validation and sanitization | N/A | No user-input handling introduced; tests feed fixed literals to private helpers. |
| High | Security | No IDOR — resource ownership validated | N/A | No resources/endpoints introduced. |
| High | Security | No SQL injection (parameterized queries) | N/A | No database access in this change. |
| High | Correctness | Logic is correct, handles edge cases | Pass | All reflection targets exist with matching signatures in `Percy/Percy.cs` (`ParseWidthsFromOptions`, `TryGetIntFromValue`, `IsResponsiveSnapshotCapture`, `CalculateDefaultHeight`, `WaitForReady`, `PayloadParser`, `setCliConfig`, `ResetInternalCaches`, `PERCY_RESPONSIVE_CAPTURE_MIN_HEIGHT`); `Cache`/`CacheItem` and the internal `PercyPlaywrightDriver(IPage)` ctor + `GetUrl()` match. Null-`IPage` invocations are safe: `WaitForReady` returns null at the disabled-preset branch before any `EvaluateSync`, and `CalculateDefaultHeight` early-returns when `PERCY_RESPONSIVE_CAPTURE_MIN_HEIGHT` is false. |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | Test code; assertions are explicit. No production error handling altered. Tests deliberately exercise the disabled/early-return branches rather than masking failures. |
| High | Correctness | No race conditions or concurrency issues | Pass | Tests mutate static `cliConfig` via `setCliConfig` but each clears it in a `finally`; no parallelism added. Static state pre-exists in `Percy.cs`. |
| Medium | Testing | New code has corresponding tests | Pass | The PR *is* the tests; the coverlet gate enforces them in CI. |
| Medium | Testing | Error paths and edge cases tested | Pass | Covers null/missing/unsupported-type options, JSON-string vs number coercion, deferUploads gating, disabled-preset readiness, already-json/null payloads, cache overwrite. |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | New class is independent; `npm test -- --no-restore` still forwards correctly (script ends with `dotnet test ...`, args append). Local coverage run produced a passing 42.4% report artifact. |
| Medium | Performance | No N+1 queries or unbounded data fetching | N/A | No data access. |
| Medium | Performance | Long-running tasks use background jobs | N/A | Not applicable to unit tests. |
| Medium | Quality | Follows existing codebase patterns | Pass | Reuses the repo's reflection-based private-static test convention (cf. `CacheTest.cs`, referenced `CorsFrameTests`) and xUnit `[Fact]`. coverlet pinned in `packages.lock.json`. |
| Medium | Quality | Changes are focused (single concern) | Pass | Scoped to coverage tooling + tests; no scope creep into production code. |
| Low | Quality | Meaningful names, no dead code | Pass | Descriptive test names; explanatory comments justify null-page safety and the config-clear trick. |
| Low | Quality | Comments explain why, not what | Pass | Comments explain *why* a null `IPage` is safe and why the class avoids the live-driver fixture — rationale, not narration. |
| Low | Quality | No unnecessary dependencies added | Pass | Single dev-only dep `coverlet.msbuild` 6.0.2 (PrivateAssets=all), the standard .NET coverage tool; legitimately required for the gate. |

## Findings

- **File:** `Percy.Test/packages.lock.json:236`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** The patch removes the trailing newline at end of file (`\ No newline at end of file` on the closing brace). Minor hygiene nit that some tooling/diffs flag.
- **Suggestion:** Restore the trailing newline. Non-blocking — the file is machine-generated by NuGet restore and will be normalized on the next lock regeneration.

- **File:** `Percy.Test/PercyLogicTest.cs:233` (also `:256`)
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** `CalculateDefaultHeight` and `WaitForReady` declare a non-nullable `IPage page` parameter in source, yet the tests invoke them via reflection with `null`. This is safe today only because both methods early-return before dereferencing `page` (the `PERCY_RESPONSIVE_CAPTURE_MIN_HEIGHT==false` and disabled-preset branches). The tests document this coupling in comments, but a future reorder of those guards could turn these into NREs surfaced as opaque `TargetInvocationException`s rather than clear failures.
- **Suggestion:** Acceptable as-is given the comments; optionally supply a lightweight `Mock<IPage>` so the tests stay valid if guard ordering changes. Non-blocking.

---

**Verdict: PASS** — test + coverage-gate only; reflection targets and signatures verified against source, null-page paths confirmed safe, coverage (42.4%) clears the 41 gate, no High/Critical issues. Only two Low hygiene nits.

**Comment by AkashBrowserStack** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/percy-playwright-dotnet/pull/236  •  **Head:** 90e9c46  •  **Reviewers:** stack:code-reviewer

## Summary
Adds a coverlet line-coverage gate (`Threshold=95`, below the achieved ~99.85%) to the `npm test` script and introduces three mock-based unit-test files (Moq + RichardSzalay.MockHttp) that exercise Percy.cs error/edge/feature-flag branches and the PercyPlaywrightDriver session/GUID logic without a live Chromium or live `@percy/cli`. The only production changes are behavior-preserving test seams: internal mirror fields for the three responsive-capture feature flags and a `DebugEnabled` mirror in Percy.cs, plus widening `PercyPlaywrightDriver.page`/`GetBrowserGuid` to `protected`, extracting a `protected virtual FetchSessionDetails()`, and adding `[InternalsVisibleTo("Percy.Test")]`. Also bumps `@percy/cli` `1.31.10-alpha.0` → stable `1.31.10` (with lockfile).

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | No secrets/tokens/keys introduced. Mock URLs/JSON only. |
| High | Security | Authentication/authorization checks present | N/A | SDK test PR; no auth surface touched. |
| High | Security | Input validation and sanitization | N/A | No new user-input boundaries; seam fields mirror existing env-derived values. |
| High | Security | No IDOR — resource ownership validated | N/A | No resource access/ownership logic. |
| High | Security | No SQL injection (parameterized queries) | N/A | No database/SQL code. |
| High | Correctness | Logic is correct, handles edge cases | Pass | Seam mirrors default to the exact existing `readonly` env values; production reads switch to mirrors with identical defaults → runtime behavior preserved. `FetchSessionDetails` wraps the unchanged `EvaluateSync` call. |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | No production error handling changed; existing debug-gated catches now read the `DebugEnabled` mirror (same default). Tests assert catch paths return null/log as designed. |
| High | Correctness | No race conditions or concurrency issues | Pass | Static-state mutating tests run under `[CollectionDefinition(... DisableParallelization = true)]` and restore originals in `Dispose`. Mocks use `Interlocked.Increment` for the shared counter. |
| Medium | Testing | New code has corresponding tests | Pass | All new seams are reached by the new tests; coverage gate enforces it. |
| Medium | Testing | Error paths and edge cases tested | Pass | success=false throws, malformed/disabled readiness, viewport/reload/reset throws, missing-widths upgrade message, HTTP 500 rethrow, resize timeout, lazy http-client init, disposed-JsonElement catch. |
| Medium | Testing | Existing tests still pass (no regressions) | Pass (not executed) | Status not independently run in this review; production behavior is unchanged by design, and the seam refactor is purely additive. CI/coverage gate is the executing authority. |
| Medium | Performance | No N+1 queries or unbounded data fetching | N/A | No data-fetch loops added (test-only). |
| Medium | Performance | Long-running tasks use background jobs | N/A | Two tests intentionally sleep ~1s (resize-timeout / sleep-time branches); acceptable for unit coverage of time-gated branches. |
| Medium | Quality | Follows existing codebase patterns | Pass | Reflection-into-private-statics mirrors the existing `CorsFrameTests` convention; `InternalsVisibleTo` and pre-existing seams (`setHttpClient`/`setCliConfig`/`ResetInternalCaches`) reused. |
| Medium | Quality | Changes are focused (single concern) | Pass | Scoped to coverage + tests + minimal seams. The `@percy/cli` alpha→stable bump is a small adjacent hygiene change, reasonable to include. |
| Low | Quality | Meaningful names, no dead code | Pass | Clear names; seam intent documented in comments. Minor: `evalDomCount`/`_ = domCount` in `ResponsivePage` is an unused out/discard kept for signature symmetry — harmless. |
| Low | Quality | Comments explain why, not what | Pass | Comments justify the seams and the Moq object-first generic-dispatch ordering — genuinely useful "why". |
| Low | Quality | No unnecessary dependencies added | Pass | `coverlet.msbuild` is a dev/test-only PrivateAssets package appropriate for the coverage gate. |

## Findings

No Critical, High, or Medium findings.

Low / informational (non-blocking):
- **File:** `package.json` — The `@percy/cli` `1.31.10-alpha.0` → `1.31.10` bump is unrelated to the coverage goal. It is reasonable hygiene (stable over alpha) but worth a one-line note in the PR description so reviewers know it is intentional.
- **File:** `Percy.Test/packages.lock.json` — Lockfile ends with "No newline at end of file"; cosmetic only.
- **File:** `Percy.Test/PercyMockedFlowTest.cs` — A couple of branch-coverage tests (`WaitForResizeCount_NeverReachesExpected...`, `CaptureResponsiveDom_SleepTimeSet...`) incur real ~1s sleeps. Fine for correctness; if suite runtime matters, these are the candidates to shorten later.

---

**Verdict: PASS** — Behavior-preserving seams with identical defaults, comprehensive mock-based coverage of error/edge/feature-flag branches, and a coverage gate floored safely below the achieved ~99.85%. No HIGH-row failures and no Critical/High findings.

**Comment by AkashBrowserStack** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/percy-playwright-dotnet/pull/236  •  **Head:** 942ce99  •  **Reviewers:** stack:code-reviewer

## Summary

Test-coverage PR for the Percy Playwright .NET SDK. It adds three new mock-based unit-test files (Moq + RichardSzalay.MockHttp) that exercise previously unreachable error/edge/feature-flag branches of `Percy.cs` and `PercyPlaywrightDriver.cs`, introduces five small behavior-preserving production "seams" (a `DebugEnabled` mirror of `DEBUG` and three `ResponsiveCapture*` mirrors of env-readonly flags in `Percy.cs`; `page` and `GetBrowserGuid` widened to `protected`/`protected virtual` plus an extracted `protected virtual FetchSessionDetails` and `[assembly: InternalsVisibleTo("Percy.Test")]` in the driver), and wires a coverlet line-coverage gate via `package.json` and `coverlet.msbuild 6.0.2`. The coverlet line floor is now 90 (lowered from 95 after a `main` merge brought in uncovered base code; ~94.8% achieved on CI). The diff also includes a routine `main` merge and a regenerated `package-lock.json` (transitive npm bumps from the official registry).

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | Test fixtures use synthetic literals (`abc123`, `real-seam-id`); no real credentials. New dep `coverlet.msbuild 6.0.2` is official NuGet, dev-only (`PrivateAssets=all`), hash-pinned. |
| High | Security | Authentication/authorization checks present | N/A | No auth surface touched. |
| High | Security | Input validation and sanitization | N/A | No new external input boundaries; tests feed controlled mock data. |
| High | Security | No IDOR — resource ownership validated | N/A | No resource-access paths changed. |
| High | Security | No SQL injection (parameterized queries) | N/A | No SQL/database code. |
| High | Correctness | Logic is correct, handles edge cases | Pass | Production seams are pure behavior-preserving refactors: mirrors initialize from the same env-derived values; `FetchSessionDetails` is a verbatim extraction of the `EvaluateSync` call from `GetSessionId`, whose caching/parsing is unchanged. |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | No new swallow/empty-catch in production code. Tests assert on existing catch branches (frame-eval throw, reload throw, viewport-reset throw) that already log at debug. |
| High | Correctness | No race conditions or concurrency issues | Pass | Tests mutate process-wide statics (cliConfig, http client, feature-flag mirrors) but run under `[Collection(... DisableParallelization=true)]` with ctor/`Dispose` save-restore. |
| Medium | Testing | New code has corresponding tests | Pass | Every new production seam is exercised; the PR is itself a test-coverage PR. |
| Medium | Testing | Error paths and edge cases tested | Pass | Covers healthcheck success=false, missing-widths throw, request-failure rethrow, malformed-readiness fallthrough, eval-throw catches, timeout loop, null-viewport fallback. |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | Seams default to prior values; CI reports green at ~94.8% line coverage. |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | None introduced. |
| Medium | Performance | Long-running tasks use background jobs | N/A | Two tests use real ~1s `Thread.Sleep`/timeout loops (genuine timing branches); serial collection adds a few wall-clock seconds. Acceptable; flagged Info. |
| Medium | Quality | Follows existing codebase patterns | Pass | Reflection-into-private-statics mirrors the existing `CorsFrameTests` convention; `InternalsVisibleTo` is the standard white-box pattern. |
| Medium | Quality | Changes are focused (single concern) | Pass | Scope is coverage + minimal seams; no unrelated production behavior change. |
| Low | Quality | Meaningful names, no dead code | Pass | One test name (`Screenshot_FromIPage_WrapsIntoDriver_AndPostsAutomateScreenshot`) overstates what it asserts, and its inline comment is verbose — cosmetic. |
| Low | Quality | Comments explain why, not what | Pass | Seam comments explain the behavior-preserving rationale; one test comment block is over-long. |
| Low | Quality | No unnecessary dependencies added | Pass | Only `coverlet.msbuild` (dev-only coverage tool) added — justified by the gate. |

## Findings

No Critical or High findings. The following Low/Info items are advisory and do not block merge:

- **File:** `package.json:4`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** Single-TFM (`net7.0`) so the line threshold is unambiguous today, but if a second TFM is ever added, coverlet checks the threshold per-framework unless `/p:ThresholdStat=total` is set. The 90 floor also leaves only a ~5-point buffer over the ~94.8% achieved.
- **Suggestion:** Optionally pin `/p:ThresholdStat=total` to future-proof the gate against multi-TFM drift.

- **File:** `Percy/Percy.cs:25` (and the three `ResponsiveCapture*` mirrors)
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** The mutable `internal static` mirrors exist solely as test seams; a future reader could mistake them for runtime-tunable config. Behavior is correct (initialized from the env-derived values; tests save-restore).
- **Suggestion:** Add a one-line XML-doc or a `_testOverride`-style naming hint. Cosmetic.

- **File:** `Percy.Test/PercyMockedFlowTest.cs:1176` (also `PercyLogicTest.cs:348,356`)
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** A few tests are existence-only / weak assertions (`*_DoesNotThrow`), and `Screenshot_FromIPage_...AndPostsAutomateScreenshot` only asserts `Assert.Null(result)` — the POST never fires because GUID reflection throws on the Moq proxy first; the name implies more than it verifies.
- **Suggestion:** Rename to match what it covers (wrapper construction) or strengthen the assertion; trim the multi-paragraph inline comment. Harmless as-is.

---

**Verdict: PASS**

**Comment by pranavz28** [comment]
+---------+--------+--------+--------+
|         | Line   | Branch | Method |
+---------+--------+--------+--------+
| Total   | 94.83% | 92.01% | 100%   |
+---------+--------+--------+--------+
| Average | 94.83% | 92.01% | 100%   |
+---------+--------+--------+--------+

currently 92%

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "private": true,
   "scripts": {
-    "test": "percy exec --testing -- dotnet test"
+    "test": "percy exec --testing -- dotnet test /p:CollectCoverage=true /p:CoverletOutputFormat=cobertura /p:ThresholdType=line /p:Threshold=90"
```
Comment (pranavz28) [inline package.json]: Keep threshold 100%

**Comment by AkashBrowserStack** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/percy-playwright-dotnet/pull/236  •  **Head:** d7358d2  •  **Reviewers:** stack:code-reviewer

## Summary
Raises PercyIO.Playwright test line coverage to 100% and bumps the coverlet line-coverage gate from 90 to 100 in package.json. Adds unit tests covering the remaining uncovered lines (Log success path; Enabled() version branches; the full ExposeClosedShadowRoots CDP flow via a Moq-able ICDPSession; WaitForReady malformed-JSON catch; GetSerializedDom corsIframes attach; Snapshot automate-throw / DOM re-inject / options-merge / no-data branches; and the driver GUID reflection). Introduces three behavior-preserving test seams and disables xUnit collection parallelization to keep the suite deterministic over Percy's process-global statics.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | Tests use literal mock JSON/guids only. |
| High | Security | Authentication/authorization checks present | N/A | No auth surface changed. |
| High | Security | Input validation and sanitization | Pass | ReadinessJsonTransform is a null-default test hook; no new external input path. |
| High | Security | No IDOR — resource ownership validated | N/A | Not applicable. |
| High | Security | No SQL injection (parameterized queries) | N/A | No DB access. |
| High | Correctness | Logic is correct, handles edge cases | Pass | Seams default to prior expressions; runtime byte-identical. |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | Existing defensive catches unchanged; tests assert their behavior. |
| High | Correctness | No race conditions or concurrency issues | Pass | DisableTestParallelization removes cross-collection static races in tests. |
| Medium | Testing | New code has corresponding tests | Pass | Every new/changed line is covered (100% line). |
| Medium | Testing | Error paths and edge cases tested | Pass | TOCTOU continue, outer/finally catches, malformed JSON all covered. |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | Full suite green on CI at Threshold=100. |
| Medium | Performance | No N+1 queries or unbounded data fetching | N/A | Not applicable. |
| Medium | Performance | Long-running tasks use background jobs | N/A | Test-only Thread.Sleep in timeout-branch tests (bounded). |
| Medium | Quality | Follows existing codebase patterns | Pass | Seams mirror the existing GetBrowserGuid/FetchSessionDetails + flag-mirror conventions. |
| Medium | Quality | Changes are focused (single concern) | Pass | Coverage + gate only. |
| Low | Quality | Meaningful names, no dead code | Pass | Clear seam names; explanatory comments. |
| Low | Quality | Comments explain why, not what | Pass | Seams documented as behavior-preserving. |
| Low | Quality | No unnecessary dependencies added | Pass | No new packages. |

## Findings

- **File:** `Percy.Test/CacheTest.cs:9`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** The assembly-level `[assembly: CollectionBehavior(DisableTestParallelization = true)]` lives in a small unrelated test file, easy to miss/delete; it also makes the per-collection `DisableParallelization` flags redundant.
- **Suggestion:** Optionally relocate to an `AssemblyInfo.cs` or next to `PercySerialTestsCollection` with a one-line note that it supersedes the per-collection flags.

- **File:** `Percy.Test/PercyMockedFlowTest.cs` (timeout-branch tests)
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** A few tests use real `Thread.Sleep` (~1-2s) to exercise timeout/sleep branches; under a now-serialized suite these add wall-clock time.
- **Suggestion:** Acceptable for a coverage milestone; revisit only if suite runtime becomes a concern (the production `WaitForResizeCount` loop isn't time-injectable).

- **File:** `package.json:4`
- **Severity:** Low (informational)
- **Reviewer:** stack:code-reviewer
- **Issue:** The gate is line-only (`ThresholdType=line`); branch/method coverage are not gated. "100%" refers to line coverage as claimed.
- **Suggestion:** No action required; accurately scoped to line coverage.

---

**Verdict: PASS** — three production seams verified behavior-preserving (each defaults to the prior expression; production runtime byte-identical), build clean, CI Test green at Threshold=100, no Critical/High findings.

File: package.json (hunk 1 above)
Comment (AkashBrowserStack) [reply package.json]: Done — the coverlet gate is now `Threshold=100` (package.json: `/p:ThresholdType=line /p:Threshold=100`). The non-browser unit suite reaches genuine 100% line coverage of the PercyIO.Playwright assembly with real tests and behavior-preserving seams — no `[ExcludeFromCodeCoverage]`, no coverage pragmas, no `[GeneratedCode]` gaming. Newly covered: the full `ExposeClosedShadowRoots` CDP flow (via a Moq-able `ICDPSession`/`NewCDPSessionAsync`), `Enabled()` version branches, `Log` success path, `WaitForReady` malformed-JSON catch, `GetSerializedDom` corsIframes attach, the `Snapshot` automate-throw/DOM-reinject/options-merge/no-data branches, and the `PercyPlaywrightDriver` GUID reflection. The CI **Test** workflow is green on the latest commit (d7358d2) with all 142 tests passing at Threshold=100.

**Review by pranavz28 — APPROVED** [review]
